docs/tutorials/data_provider.md

```diff
@@ -61,3 +61,169 @@ impl AdditiveIdentity {
+                // Note: Cloning a DataPayload is usually cheap, and it is necessary in order to
+                // convert the short-lived cache object into one we can return.
+                return any_res.downcast_cloned();
+            }
+        }
+        // Release the lock to invoke the inner provider
+        let response = self.provider.load(req)?;
+        let owned_cache_key = CacheKeyWrap(CacheKey(M::KEY, Cow::Owned(req.locale.clone())));
+        // Second lock: cache storage
+        self.cache.lock()
+            .unwrap()
+            .get_or_insert(owned_cache_key, || response.wrap_into_any_response())
+            .downcast_cloned()
+    }
+}
+
+// Usage example:
+let provider = icu_testdata::buffer();
+let lru_capacity = 100usize.try_into().unwrap();
+let provider = LruDataCache {
+    cache: Mutex::new(LruCache::new(lru_capacity)),
+    provider: provider.as_deserializing(),
+};
+
+// The cache starts empty:
+assert_eq!(provider.cache.lock().unwrap().len(), 0);
+
+assert_eq!(
+    "こんにちは世界",
+    // Note: It is necessary to use `try_new_unstable` with LruDataCache.
+    HelloWorldFormatter::try_new_unstable(
+        &provider,
+        &locale!("ja").into()
+    )
+    .unwrap()
+    .format_to_string()
+);
+
+// One item in the cache:
+assert_eq!(provider.cache.lock().unwrap().len(), 1);
+
+assert_eq!(
+    "ওহে বিশ্ব",
+    HelloWorldFormatter::try_new_unstable(
+        &provider,
+        &locale!("bn").into()
+    )
+    .unwrap()
+    .format_to_string()
+);
+
+// Two items in the cache:
+assert_eq!(provider.cache.lock().unwrap().len(), 2);
+
+assert_eq!(
+    "こんにちは世界",
+    HelloWorldFormatter::try_new_unstable(
+        &provider,
+        &locale!("ja").into()
+    )
+    .unwrap()
+    .format_to_string()
+);
+
+// Still only two items in the cache, since we re-requested "ja" data:
+assert_eq!(provider.cache.lock().unwrap().len(), 2);
+```
+
+## Overwriting Specific Data Items
+
+ICU4X's explicit data pipeline allows for specific data entries to be overwritten in order to customize the output or comply with policy.
+
+The following example illustrates how to overwrite the decimal separators for a region.
+
+```rust
+use icu::decimal::FixedDecimalFormatter;
+use icu_provider::prelude::*;
+use icu::locid::locale;
+use icu::locid::subtags_region as region;
+use std::borrow::Cow;
+use tinystr::tinystr;
+
+pub struct CustomDecimalSymbolsProvider<P>(P);
+
+impl<P> AnyProvider for CustomDecimalSymbolsProvider<P>
+where
+    P: AnyProvider
+{
+    fn load_any(&self, key: DataKey, req: DataRequest) -> Result<AnyResponse, DataError> {
+        use icu::decimal::provider::DecimalSymbolsV1Marker;
+        let mut any_res = self.0.load_any(key, req)?;
+        if key == DecimalSymbolsV1Marker::KEY && req.locale.region() == Some(region!("CH")) {
+            let mut res: DataResponse<DecimalSymbolsV1Marker> = any_res.downcast()?;
+            if let Some(payload) = &mut res.payload.as_mut() {
```

Comment:
I was wondering what the error case here is, but there doesn't seem to be one. The design decision to make payload an `Option<DataPayload>` doesn't seem to have paid off, it's never `None`. Maybe `unwrap` here so readers don't have to figure this out themselves.

We should consider making the payload non-optional for 2.0 if there's still no use case for an absent payload by then.

Reply:
https://github.com/unicode-org/icu4x/issues/2956